src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs

```diff
@@ -59,10 +74,32 @@ internal override BoundStatement BindForEachParts(DiagnosticBag diagnostics, Bin
             return result;
         }
 
+        /// <summary>
+        /// Like BindForEachParts, but only bind the deconstruction part of the foreach, for purpose of inferring the types of the declared locals.
+        /// </summary>
+        internal override void BindForEachDeconstruction(DiagnosticBag diagnostics, Binder originalBinder)
+        {
+            // Use the right binder to avoid seeing iteration variable
+            BoundExpression collectionExpr = originalBinder.GetBinder(_syntax.Expression).BindValue(_syntax.Expression, diagnostics, BindValueKind.RValue);
+
+            ForEachEnumeratorInfo.Builder builder = new ForEachEnumeratorInfo.Builder();
+            TypeSymbol inferredType;
+            bool hasErrors = !GetEnumeratorInfoAndInferCollectionElementType(ref builder, ref collectionExpr, diagnostics, out inferredType);
+
+            VariableDeclarationSyntax variables = _syntax.DeconstructionVariables;
+            var valuePlaceholder = new BoundDeconstructValuePlaceholder(_syntax.Expression, inferredType ?? CreateErrorType("var"));
```

Comment:
The expression `inferredType ?? CreateErrorType("var")` is used several times in `ForEachLoopBinder`. Consider extracting a helper method.